### PR TITLE
Redirect: redirect to permalink instead of editor for logged out users

### LINKED ISSUE #53017: (Revisiting) Site Settings: Redirect calypso links for site visitors, and show a proper link to the post.
**Request:** If a logged-out or non-site user clicks a Calypso edit link of a published post, please redirect to the post itself to its proper URL, with a message at the top inviting the site owner to log in. (A revisit of [this issue](https://github.com/Automattic/wp-calypso/issues/4657).)

Please track this redirect as well as how often someone clicks to log in / switch accounts. If most visitors are not clicking log-in links, we should probably make it easier for site owners to find the proper permalink. 

**Why**
We've visited this issue before [here](https://github.com/Automattic/wp-calypso/issues/4657)  but I still frequently see people sharing links to Calypso editor pages. 

They expect their friends and family and customers to see this:

Screen Shot 2022-03-07 at 10 27 27 AM

Instead they get this:
Screen Shot 2022-10-12 at 12 21 26 PM

**How**

_What we could do instead:_
- If the site is public and the post link is published, we assume site visitors are supposed to see the post, not the edit page.
- We redirect to the post
- We add a SMALL banner at the top in case the visitor is in fact a site member trying to edit the post, inviting them to log in. That could redirect back to the post edit page if we want to.

Also since I suspect folks are sharing that link from the preview page, making a more obvious share/copy link icon there might be smart. I have made a separate request for that: https://github.com/Automattic/wp-calypso/issues/53019

**Additional context**
Just a few examples of people sharing a Calypso link. There are more on Facebook, Twitter, in emails, etc.  And I suspect if we tracked how often site visitors are clicking those links, we'd find it happens all the time.
https://wordpress.com/forums?s=https%3A%2F%2Fwordpress.com%2Fpost%2F

### PR DESCRIPTION
**Pull request by enejb**
Alternative to #61925, 
Fixes https://github.com/Automattic/wp-calypso/issues/53017

This PR redirects users of the post editor to a new endpoint defined in D76631-code that then decides to either send the user back to the login screen ( if the post is not published ) or to the permalink.

#### Changes proposed in this Pull Request

Based on the feedback from @tyxla on https://github.com/Automattic/wp-calypso/pull/61925. 
This PR does the following:

1. It make the `editor` section be 'enableLoggedOut` so that we can be really specific as to what should happen if a loggout user visits it. 

2. If you are visit any section other then 
`'/edit/:customPostType/:site/:post?', '/post/:site/:post?' or '/page/:site/:post?' we send the user to loggin if they haven
't already. 

For logged-in users, the experience should stay the same.

Related to D76631-code

#### Testing instructions

    Apply D76631-code to your sandbox and sandbox public-api.wordpress.com to point to your sandbox.
    Copy the url of a published ( page, post, portfolio and testimony ) in the editor.
    In an incognito ( logged out view ) visit the editor page that you published.
    It should take you to the permalink of that post, page, portfolio and testimony instead of the log-in form.
    Note that if you are logged in that you should see the editor.

### CODE REVIEW

File: client/gutenberg/editor/index.js
Comment (tyxla) [inline client/gutenberg/editor/index.js]: Do you mind moving this one to `client/gutenberg/editor/controller.js`? That's where the rest of the similar controller functions live.

File: client/gutenberg/editor/index.js
Comment (tyxla) [inline client/gutenberg/editor/index.js]: Seems like this will only work with immediate login. Is that intended? 

Also, do we need any additional checks - like whether it exists?

Comment (tyxla) [inline client/gutenberg/editor/index.js]: Seems like this will only work with immediate login. Is that intended?

File: client/gutenberg/editor/index.js
Comment (tyxla) [inline client/gutenberg/editor/index.js]: This comment seems to be off; we don't really reload Calypso here, we actually redirect to an endpoint that will redirect us somewhere itself.

Comment (tyxla) [inline client/gutenberg/editor/index.js]: This comment seems to be off; we don't really reload Calypso here, we actually redirect to an endpoint that will redirect us somewhere itself.

**Review by tyxla — COMMENTED** [review]
I dig this approach @enejb 👍 

I have a few questions, but I think it should be pretty close.

File: client/gutenberg/editor/index.js
Comment (enejb) [reply client/gutenberg/editor/index.js]: Good call.

File: client/gutenberg/editor/index.js
Comment (enejb) [reply client/gutenberg/editor/index.js]: I am not really sure if this is the right approach here. Ideally, we redirect the user to the correct locale. 
`getImmediateLoginLocale( state );` call is something that is done in the `isUserLoggedIn` function. So I kept it but I am not sure how to use it.

Comment (enejb) [reply client/gutenberg/editor/index.js]: I am not sure. Should we check the users browser locale in order to send them to the correct login in the end?

File: client/gutenberg/editor/index.js
Comment (enejb) [reply client/gutenberg/editor/index.js]: Good call.

File: client/gutenberg/editor/index.js
Comment (enejb) [reply client/gutenberg/editor/index.js]: I removed the call since I don't think its applicable.

File: client/gutenberg/editor/index.js
Comment (enejb) [reply client/gutenberg/editor/index.js]: Done

File: client/gutenberg/editor/index.js
Comment (enejb) [reply client/gutenberg/editor/index.js]: I updated the comment

**Comment by enejb** [comment]
Thanks for all the feedback Marin! 🙇

File: client/gutenberg/editor/controller.js
Comment (tyxla) [inline client/gutenberg/editor/controller.js]: I wonder if we could also add another check and do this only for certain post types. We're doing it in the endpoint already, but we could also do it here and save an extra check or redirect if the post type isn't among the supported ones.

**Review by tyxla — COMMENTED** [review]
Code looks good to me in general 👍  Nice work.

I've left some feedback in the accompanying diff, and asked a question if we could further tighten up the logic before redirecting.

I'm not approving, because some of my comments in the diff may require further changes in this PR if you agree with the sentiment.

**Comment by supernovia** [comment]
👋  Just ran into another case like this today @enejb  - let me know if there's anything we can do to help from our side 🎸

**Comment by enejb** [comment]
@supernovia Sorry this is still on my list of things to continue working on. Its just has a lower priority. I am hoping to get to make progress on this in the next few weeks.

**Comment by supernovia** [comment]
> @supernovia Sorry this is still on my list of things to continue working on. Its just has a lower priority. I am hoping to get to make progress on this in the next few weeks.

I noticed this at "Squash and Merge now" -- checking in to see if there's anything else we can do!

**Comment by enejb** [comment]
@tyxla  I updated  D76631-code as well as rebased this code and fixes all the issues that you requested. 
Can you take another look when you have a moment? 
Thank you!

File: client/gutenberg/editor/controller.js
Comment (tyxla) [inline client/gutenberg/editor/controller.js]: This is only the case if the redirect conditions aren't fulfilled. If we end up redirecting, it will return `undefined`.

File: client/gutenberg/editor/index.js
Comment (tyxla) [inline client/gutenberg/editor/index.js]: cc @mattwiebe: does this need to be further tested with regards to the recent site editor un-framing in p7DVsv-h8i-p2 ?

File: client/gutenberg/editor/controller.js
Comment (tyxla) [inline client/gutenberg/editor/controller.js:318]: This bit is a bit fragile and dependable on the exact format of the paths. Can we add a comment to clarify with some examples the expected paths this works with?

**Review by tyxla — COMMENTED** [review]
This looks mostly good to me, @enejb! I've left a question in the corresponding diff, plus a couple of minor suggestions and a ping to confirm this is good for the site editor.

Great work IMHO! 👍 

Going to need a rebase btw.

File: client/gutenberg/editor/index.js
Comment (mattwiebe) [reply client/gutenberg/editor/index.js]: @tyxla @enejb I don't think that what this PR is doing should have anything to do with the Site Editor - this is definitely out of date with the [new routing](https://github.com/Automattic/wp-calypso/blob/fd388ceede5996f6480bf6cc41eaf4e8b313d0d1/client/gutenberg/editor/index.js#L7) which is just

```js
page( '/site-editor/:site?', siteSelection, redirectSiteEditor );
```

And with https://github.com/Automattic/jetpack/pull/30072 due to land next week, the `/site-editor/SITE` route will cease most usage anyway, it'll just stay there for back compat and for whatever places hardcode the `/site-editor` route in Calypso.

File: client/gutenberg/editor/controller.js
Comment (enejb) [reply client/gutenberg/editor/controller.js:318]: I added a comment about the possible routes

File: client/gutenberg/editor/controller.js
Comment (tyxla) [inline client/gutenberg/editor/controller.js]: Something just occurred to me: will this be available for all kinds of sites, like Jetpack/Atomic? Do we need any special handling for those site types?

**Review by tyxla — COMMENTED** [review]
I just have one question before this is ready to go. 👍 

Also, I think it should still be rebased and tested finally since there are some merge conficts.

File: client/gutenberg/editor/controller.js
Comment (tyxla) [inline client/gutenberg/editor/controller.js]: Does `getSiteFragment()` properly support sites installed in subdirectories? I didn't see that in its unit tests.

Comment (tyxla) [inline client/gutenberg/editor/controller.js]: I just realized that we don't really verify that the site fragment will be properly read as a proper site by the REST API. If this `window.location` redirects to a 404 it would be a pretty bad user experience.

File: client/gutenberg/editor/index.js
Comment (tyxla) [reply client/gutenberg/editor/index.js]: Thank you for confirming @mattwiebe 🙌

File: client/gutenberg/editor/controller.js
Comment (enejb) [reply client/gutenberg/editor/controller.js]: A 404 only happends when the site doesn't exits. 
Which should never happen though regular use. A user would have to go and manually change the url.

Comment (enejb) [reply client/gutenberg/editor/controller.js]: In my testing sites with subdirectories work as expected here.

Comment (enejb) [reply client/gutenberg/editor/controller.js]: Yes this should work for all sites as long as jetpack sync works. 
If jetpack sync is not working then we might be falling back on the user being sent to the (wordpress.com) login screen or a 404 page on the site. 

I would say this is a rare case anyways. (atomic and jetpack site) since they currently get redirected to the wp-admin side of things and this redirect fix is doesn't work for that case at all.

**Review by tyxla — APPROVED** [review]
From my perspective, this should be good to go once rebased (and the related diff has been merged).

Thanks @enejb 🚀

**Comment by simison** [comment]
Neat! 

Was this pre-existing behaviour, or do we now redirect to _default site's editor_ without warning if I login with different user than what has access to the site I originally linked to?

1) Open https://wordpress.com/page/cicerosletters.wordpress.com/38
2) See login page, login with user without access to above site
3) See https://wordpress.com/page/test1162.home.blog
4) I might still believe I'm looking at first site but actually I'm gonna publish on another site.

As a future improvement, we could improve the copy at the login page so that it's not just a blunt redirection. I like the example from Substack:

Screenshot 2023-04-26 at 12 10 46

vs ours:

Screenshot 2023-04-28 at 13 00 42

**Comment by enejb** [comment]
This is a great suggestion for improvement @simison! 

If the site is private or the post is not public the flow is the same as before.
If the post is public and the user is logged out we send them to the permalink.

So in this PR we only tackled one of the cases. (And hopefully the most common case) 
We get asked about this in the forums. Such as https://wordpress.com/forums/topic/link-https-wordpress-com-post-drkorner-wordpress-com-423-doesnt-load/

**Comment by tyxla** [comment]
Yup, what Enej said (also, Enej, that's not how AFK is supposed to be done 😅 ).

@simison I think your suggestion can be documented in another issue since it's not 100% related and was still an issue before this change. Worth having though, which is why I'd recommend opening it as an issue.

**Pull request merged**
